simulation/amaru-sim/src/simulator/run.rs

```diff
@@ -315,3 +284,72 @@ impl BlockFetcher for FakeBlockFetcher {
         Ok(vec![])
     }
 }
+
+/// This implementation of ForwardEventListener sends the received events a Sender that can collect them
+/// to check them later.
+///
+/// A message id is maintained for each received event and each message is duplicated to the number of downstream peers.
+#[derive(Clone, Debug)]
+pub struct MockForwardEventListener {
+    node_id: String,
+    number_of_downstream_peers: u8,
+    sender: mpsc::Sender<Envelope<ChainSyncMessage>>,
+    msg_id: Arc<Mutex<u64>>,
+}
+
+impl MockForwardEventListener {
+    pub fn new(
+        node_id: String,
+        number_of_downstream_peers: u8,
+        sender: mpsc::Sender<Envelope<ChainSyncMessage>>,
+    ) -> Self {
+        Self {
+            node_id,
+            number_of_downstream_peers,
+            msg_id: Arc::new(Mutex::new(0)),
+            sender,
+        }
+    }
+}
+
+#[async_trait]
+impl ForwardEventListener for MockForwardEventListener {
+    async fn send(&self, event: ForwardEvent) -> anyhow::Result<()> {
+        fn message(event: &ForwardEvent, msg_id: u64) -> ChainSyncMessage {
+            match event {
+                ForwardEvent::Forward(header) => ChainSyncMessage::Fwd {
+                    msg_id,
+                    slot: header.point().slot_or_default(),
+                    hash: Bytes {
+                        bytes: header.hash().as_slice().to_vec(),
+                    },
+                    header: Bytes {
+                        bytes: to_cbor(&header),
+                    },
+                },
+                ForwardEvent::Backward(tip) => ChainSyncMessage::Bck {
+                    msg_id,
+                    slot: tip.point().slot_or_default(),
+                    hash: Bytes {
+                        bytes: tip.hash().as_slice().to_vec(),
+                    },
+                },
+            }
+        }
+
+        for i in 1..=self.number_of_downstream_peers {
+            let envelope = {
+                let mut msg_id = self.msg_id.lock().await;
```

Comment:
this would then become `self.msg_id.fetch_add(1, Ordering::Relaxed)`; if consecutive numbers are important, add `self.number_of_downstream_peers` before the loop and add `i - 1` to the result

Reply:
Done.